File: algorithms/[740]删除并获得点数/solution.ts
// topics = ["动态规划"]

/**
 * DP
 * time O(n + m), space O(m), n 为数组长度, m 为元素的最大值
 */
function deleteAndEarn(nums: number[]): number {
  const maxVal = Math.max(...nums);
  const sum = new Array<number>(maxVal + 1).fill(0);
  for (const num of nums) {
    sum[num] += num;
  }

  // 转换为「打家劫舍」问题
  return rob(sum);
}

function rob(nums: number[]): number {
  const n = nums.length;
  if (n == 0) {
    return 0;
  }
  if (n == 1) {
    return nums[0];
  }

  let first = nums[0];
  let second = Math.max(nums[0], nums[1]);

  for (let i = 2; i < n; i += 1) {
    [first, second] = [second, Math.max(first + nums[i], second)];
  }

  return second;
}

File: algorithms/[740]删除并获得点数/solution.py
# topics = ["动态规划"]

from typing import List
from collections import Counter


class Solution:
    def deleteAndEarn(self, nums: List[int]) -> int:
        """
        近似题 198.打家劫舍
        DP
        time O(nlogn), space O(n), n 为数组长度
        """
        cnt = Counter(nums)
        _nums = sorted(cnt.keys())
        n = len(_nums)

        # dp[i] 表示前 i 个数的最大点数
        dp = [0] * (n + 1)
        dp[1] = _nums[0] * cnt[_nums[0]]

        for i in range(2, n + 1):
            total = _nums[i - 1] * cnt[_nums[i - 1]]
            if _nums[i - 1] - _nums[i - 2] == 1:
                dp[i] = max(dp[i - 1], dp[i - 2] + total)
            else:
                dp[i] = dp[i - 1] + total

        return dp[n]
